### PR TITLE
Cherry pick support for Complex Number type

### PR DESCRIPTION
**Pull request by atris**

### CODE REVIEW

**Comment by hlinnaka** [comment]
Why do we want a complex number datatype?

File: src/test/regress/bugbuster/known_good_schedule [hunk 1]
```diff
@@ -25,4 +25,5 @@ test: bkup_pg gpmgmt1 cursor metadata_track
 test: memory_quota
 test: TestPSQL
 test: opperf
+test: complex
```
Comment (hlinnaka) [inline src/test/regress/bugbuster/known_good_schedule]: Let's not add new tests to bugbuster, but it in the regular installcheck-good regression suite instead

Comment (yydzero) [inline src/test/regress/bugbuster/known_good_schedule]: +1 for ICG

**Comment by ivannovick** [comment]
@hlinnaka we use complex number types in some data science use cases for Oil & Gas industry, i think its fast fourier transforms or some other calcs that this enables.

File: src/test/regress/sql/complex.sql [hunk 2]
```diff
@@ -0,0 +1,388 @@
+--
```
Comment (yydzero) [inline src/test/regress/sql/complex.sql]: Missed expected out file? Did not see it in this commit.

**Comment by yydzero** [comment]
This feature is a request from PM, @ivannovick

**Comment by mkiyama** [comment]
Do you have a link to documentation for this feature?

File: src/test/regress/bugbuster/known_good_schedule (hunk 1 above)
Comment (vraghavan78) [inline src/test/regress/bugbuster/known_good_schedule]: +1

**Comment by danielgustafsson** [comment]
For backports from upstream I would like the commit message(s) of the upstream commit(s) that were included in the cherry-pick commit message.

Also, since this is labelled part 1, can you briefly describe what will be in part 2 (and 3..n in case more than two parts are expected)?

**Comment by atris** [comment]
@danielgustafsson This is not an upstream cherry pick.

Sorry for the incorrect commit message in my branch. This was supposed to be done in parts but then I decided to do the whole thing at once

**Comment by danielgustafsson** [comment]
Right, I was fooling myself then when scanning over the diff and seeing the copyright headers. When not being a backport the copyright headers are wrong in that they should at least have a "Portions copyright Greenplum" line I would assume? Also, if we aren't bound by keeping in sync with upstream I would argue that we don't need the `HAVE_BUGGY_IRIX_STRTOD` block since we don't support building on IRIX.

File: src/test/regress/sql/complex.sql (hunk 2 above)
Comment (atris) [inline src/test/regress/sql/complex.sql]: Good catch. Local gitignore twerks. Updated the PR

**Comment by atris** [comment]
@danielgustafsson which headers are you referring to,please?

File: src/backend/utils/adt/complex_type.c [hunk 3]
```diff
@@ -0,0 +1,1279 @@
...
+ * complex.c
+ *	  Core I/O functions for the built-in complex type.
+ *
+ * Portions Copyright (c) 1996-2009, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ *
+ * IDENTIFICATION
+ *
+ *-------------------------------------------------------------------------
```
Comment (danielgustafsson) [inline src/backend/utils/adt/complex_type.c]: @atris this, and any other like it, are the file headers I'm referring to. If we have included parts of upstream code the copyrights here should remain but if we have written parts there should be a "Portions Copyright (c) 2016 Greenplum" (or Pivotal or whatever it is we usually put, see other files in the repo for examples).

The IDENTIFICATION line is also an upstream artefact that should be completed with the path to the file, see the header in `src/interfaces/libpq/fe-auth.c` for an example.

File: src/test/regress/bugbuster/known_good_schedule (hunk 1 above)
Comment (atris) [inline src/test/regress/bugbuster/known_good_schedule]: Fixed

File: src/backend/utils/adt/complex_type.c (hunk 3 above)
Comment (atris) [inline src/backend/utils/adt/complex_type.c]: Let me look into this,thanks!

Comment (atris) [inline src/backend/utils/adt/complex_type.c]: @ivannovick Can you please comment on the copyright part?

Comment (ivannovick) [inline src/backend/utils/adt/complex_type.c]: I think @danielgustafsson's advise is correct... we use Pivotal Software Inc now (rather than Greenplum).

Comment (atris) [inline src/backend/utils/adt/complex_type.c]: fixed

File: src/include/catalog/pg_proc.h [hunk 4]
```diff
@@ -4228,6 +4228,191 @@ DESCR("read the error log for the specified external table");
...
+DATA(insert OID = 3466 ( sqrt  PGNSP PGUID 12 1 0 0 f f t f i 1 0 193 f "193" _null_ _null_ _null_ _null_ complex_sqrt _null_ _null_ c ));
+DESCR("squre root");
+
+/* cbrt(complex) => complex */ 
+DATA(insert OID = 3467 ( cbrt  PGNSP PGUID 12 1 0 0 f f t f i 1 0 193 f "193" _null_ _null_ _null_ _null_ complex_cbrt _null_ _null_ c ));
+DESCR("cube root");
+
+/* numeric2point("numeric") => complex */ 
+DATA(insert OID = 3468 ( numeric2point  PGNSP PGUID 12 1 0 0 f f t f i 1 0 193 f "1700" _null_ _null_ _null_ _null_ numeric2complex _null_ _null_ c ));
+DESCR("(internal) type cast from numeric to complex");
```
Comment (danielgustafsson) [inline src/include/catalog/pg_proc.h]: Why is it called `numeric2point()` when it returns a complex and not in fact the point data type?

**Comment by danielgustafsson** [comment]
The headers look good now, thanks.

**Comment by ascherbaum-pivotal** [comment]
@atris Any update on this patch? It's idle for almost a month.

**Comment by atris** [comment]
This PR is outdated. I will close this and submit a new PR for the new patch

**Pull request closed without merging**